go/upgrade/upgrade.go

```diff
@@ -279,8 +279,29 @@ func (u *upgradeManager) ConsensusUpgrade(privateCtx interface{}, currentEpoch b
 			if err := u.flushDescriptorLocked(); err != nil {
 				return err
 			}
-			u.shouldStop = true // Ensure we really stop before proceeding.
-			return api.ErrStopForUpgrade
+			// Check if we can proceed in place (e.g. without a restart).
+			u.shouldStop = func() bool {
+				if err := pu.Descriptor.EnsureCompatible(); err != nil {
```

Comment:
Not part of this PR, but it looks like we can change the descriptor after it was proposed and accepted. Do we check that the fingerprint of the upgrade methods stayed the same? :thinking:

Reply:
What do you mean "change the descriptor"? Upgrade descriptors are local (e.g. can be submitted even without governance as that is the mechanism that was present before there was governance support) and governance can just push them to all nodes.

In theory node operators could cancel these locally, but there is a fallback mechanism for this case where the consensus governance service would trigger a descriptor reinstall and node shutdown in case the upgrade did not happen. Without it there would be risk of state corruption.